<a href="https://colab.research.google.com/github/nimratmann/datasci_2_manipulation/blob/main/HHA507Assignment2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. Data Cleaning with Pandas:

## Loading Pandas Packages

In [ ]:
import pandas as pd

##Loading Healthcare Dataset

In [ ]:
healthcare_data_cleaning = pd.read_csv('https://raw.githubusercontent.com/nimratmann/datasci_2_manipulation/main/data/healthcare%20data%20cleaning.csv')
healthcare_data_cleaning

Patient Age  Gender  City of Residence State of Residence  \
0               45   Other         West Brian           Kentucky   
1               25  Female   East Jocelynfurt           Colorado   
2               51   Other  South Lindseyland               Ohio   
3               18  Female         Taylorfort       Pennsylvania   
4               48  Female          Birdmouth            Montana   
...            ...     ...                ...                ...   
104995          74   Other       West Anthony           Oklahoma   
104996          37    Male          Mckeetown               Utah   
104997          51  Female   Lake Lisachester           Nebraska   
104998     missing  Female       Heidiborough            missing   
104999          45  Female      South Jeffrey      Massachusetts   

       Has Insurance Visited Last Month Payment Method Preferred Doctor  \
0                Yes                Yes      Insurance     Dr. Williams   
1                 No                Yes           Card      Dr. Johnson   
2                Yes                 No           Cash     Dr. Williams   
3                Yes            missing           Cash     Dr. Williams   
4                 No                 No      Insurance        Dr. Brown   
...              ...                ...            ...              ...   
104995           Yes                 No           Card     Dr. Williams   
104996           Yes                 No           Cash        Dr. Jones   
104997           Yes                Yes      Insurance     Dr. Williams   
104998           Yes                 No           Cash        Dr. Smith   
104999            No                Yes           Cash        Dr. Brown   

       Disease Diagnosed Medication Prescribed Type of Appointment  \
0                    Flu                 Med_E             General   
1                    Flu                 Med_B          Specialist   
2               Covid-19                 Med_D             General   
3                    Flu                 Med_A             General   
4                   None                 Med_C          Specialist   
...                  ...                   ...                 ...   
104995          Covid-19                 Med_E          Specialist   
104996          Covid-19                 Med_C             General   
104997           Allergy                 Med_D           Follow-Up   
104998          Covid-19                 Med_A           Follow-Up   
104999           Allergy                 Med_D             General   

       Average Heart Rate Average BP Height (in cm) Weight (in kg)  \
0                      66        111            183             70   
1                      59    missing            174             58   
2                      79        119            161             56   
3                      99        115            171             52   
4                      93         89            186             87   
...                   ...        ...            ...            ...   
104995                 65        112            169             61   
104996            missing        111            165             67   
104997                 89        118            179             71   
104998                 74        123            167             75   
104999                 77        140            167             62   

           Payment Due ($) Last Visit (days ago) Visit Duration (mins)  \
0        65.45721578126224                   193                    92   
1       430.68365678679174                   195               missing   
2        315.0709305262176               missing                    37   
3        320.2998987723972                   175                   108   
4        264.2147372473319                   188                   113   
...                    ...                   ...                   ...   
104995  309.16444606680665                   183                    10   
104996         

## Identifying missing values in the dataset
### Changing "missing" to "NaN"

In [ ]:
healthcare_data_cleaning.replace('missing', pd.np.nan, inplace=True)
healthcare_data_cleaning

<ipython-input-6-724cc7559546>:1: FutureWarning: The pandas.np module is deprecated and will be removed from pandas in a future version. Import numpy directly instead.
  healthcare_data_cleaning.replace('missing', pd.np.nan, inplace=True)


Patient Age  Gender  City of Residence State of Residence  \
0               45   Other         West Brian           Kentucky   
1               25  Female   East Jocelynfurt           Colorado   
2               51   Other  South Lindseyland               Ohio   
3               18  Female         Taylorfort       Pennsylvania   
4               48  Female          Birdmouth            Montana   
...            ...     ...                ...                ...   
104995          74   Other       West Anthony           Oklahoma   
104996          37    Male          Mckeetown               Utah   
104997          51  Female   Lake Lisachester           Nebraska   
104998         NaN  Female       Heidiborough                NaN   
104999          45  Female      South Jeffrey      Massachusetts   

       Has Insurance Visited Last Month Payment Method Preferred Doctor  \
0                Yes                Yes      Insurance     Dr. Williams   
1                 No                Yes           Card      Dr. Johnson   
2                Yes                 No           Cash     Dr. Williams   
3                Yes                NaN           Cash     Dr. Williams   
4                 No                 No      Insurance        Dr. Brown   
...              ...                ...            ...              ...   
104995           Yes                 No           Card     Dr. Williams   
104996           Yes                 No           Cash        Dr. Jones   
104997           Yes                Yes      Insurance     Dr. Williams   
104998           Yes                 No           Cash        Dr. Smith   
104999            No                Yes           Cash        Dr. Brown   

       Disease Diagnosed Medication Prescribed Type of Appointment  \
0                    Flu                 Med_E             General   
1                    Flu                 Med_B          Specialist   
2               Covid-19                 Med_D             General   
3                    Flu                 Med_A             General   
4                   None                 Med_C          Specialist   
...                  ...                   ...                 ...   
104995          Covid-19                 Med_E          Specialist   
104996          Covid-19                 Med_C             General   
104997           Allergy                 Med_D           Follow-Up   
104998          Covid-19                 Med_A           Follow-Up   
104999           Allergy                 Med_D             General   

       Average Heart Rate Average BP Height (in cm) Weight (in kg)  \
0                      66        111            183             70   
1                      59        NaN            174             58   
2                      79        119            161             56   
3                      99        115            171             52   
4                      93         89            186             87   
...                   ...        ...            ...            ...   
104995                 65        112            169             61   
104996                NaN        111            165             67   
104997                 89        118            179             71   
104998                 74        123            167             75   
104999                 77        140            167             62   

           Payment Due ($) Last Visit (days ago) Visit Duration (mins)  \
0        65.45721578126224                   193                    92   
1       430.68365678679174                   195                   NaN   
2        315.0709305262176                   NaN                    37   
3        320.2998987723972                   175                   108   
4        264.2147372473319                   188                   113   
...                    ...                   ...                   ...   
104995  309.16444606680665                   183                    10   
104996         

## Removing duplicate rows
### There are no duplicate columns to remove

In [ ]:
duplicate_rows = healthcare_data_cleaning.duplicated()
healthcare_data_cleaning.drop_duplicates(inplace=True)
healthcare_data_cleaning

patientage  gender    cityofresidence stateofresidence hasinsurance  \
0              45   Other         West Brian         Kentucky          Yes   
1              25  Female   East Jocelynfurt         Colorado           No   
2              51   Other  South Lindseyland             Ohio          Yes   
3              18  Female         Taylorfort     Pennsylvania          Yes   
4              48  Female          Birdmouth          Montana           No   
...           ...     ...                ...              ...          ...   
104994         53  Female      Michelleburgh     South Dakota           No   
104995         74   Other       West Anthony         Oklahoma          Yes   
104996         37    Male          Mckeetown             Utah          Yes   
104997         51  Female   Lake Lisachester         Nebraska          Yes   
104998        NaN  Female       Heidiborough              NaN          Yes   

       visitedlastmonth paymentmethod preferreddoctor diseasediagnosed  \
0                   Yes     Insurance    Dr. Williams              Flu   
1                   Yes          Card     Dr. Johnson              Flu   
2                    No          Cash    Dr. Williams         Covid-19   
3                   NaN          Cash    Dr. Williams              Flu   
4                    No     Insurance       Dr. Brown             None   
...                 ...           ...             ...              ...   
104994               No          Card    Dr. Williams              Flu   
104995               No          Card    Dr. Williams         Covid-19   
104996               No          Cash       Dr. Jones         Covid-19   
104997              Yes     Insurance    Dr. Williams          Allergy   
104998               No          Cash       Dr. Smith         Covid-19   

       medicationprescribed typeofappointment averageheartrate averagebp  \
0                     Med_E           General               66       111   
1                     Med_B        Specialist               59       NaN   
2                     Med_D           General               79       119   
3                     Med_A           General               99       115   
4                     Med_C        Specialist               93        89   
...                     ...               ...              ...       ...   
104994                Med_B         Emergency               68       109   
104995                Med_E        Specialist               65       112   
104996                Med_C           General              NaN       111   
104997                Med_D         Follow-Up               89       118   
104998                Med_A         Follow-Up               74       123   

       heightincm weightinkg          paymentdue lastvisitdaysago  \
0             183         70   65.45721578126224              193   
1             174         58  430.68365678679174              195   
2             161         56   315.0709305262176              NaN   
3             171         52   320.2998987723972              175   
4             186         87   264.2147372473319              188   
...           ...        ...                 ...              ...   
104994        172         75   488.0316041186948              167   
104995        169         61  309.16444606680665              183   
104996        165         67                 NaN              191   
104997        179         71   320.8304180033907              172   
104998        167         75   5.341231420036174              180   

       visitdurationmins numberoftests    prescriptioncost  
0                     92             5   15.71751735532889  
1                    NaN             4   80.78647284463952  
2                     37             2    64.3921393009105  
3                    108             2   8.872859215315316  
4                    113             3   77.48311304645532  
...                  ...           ...                 ...  
104994      

## Cleaning column names
### Removing White Space and Special Characters


In [ ]:
import re

def clean_column_names(healthcare_data_cleaning):
    def clean_name(name):
        cleaned_name = re.sub(r'[^a-zA-Z0-9]', '', name)
        return cleaned_name.lower()


    healthcare_data_cleaning.columns = [clean_name(col) for col in healthcare_data_cleaning.columns]
    return healthcare_data_cleaning

healthcare_data_cleaning = clean_column_names(healthcare_data_cleaning)

print(healthcare_data_cleaning)

       patientage  gender    cityofresidence stateofresidence hasinsurance  \
0              45   Other         West Brian         Kentucky          Yes   
1              25  Female   East Jocelynfurt         Colorado           No   
2              51   Other  South Lindseyland             Ohio          Yes   
3              18  Female         Taylorfort     Pennsylvania          Yes   
4              48  Female          Birdmouth          Montana           No   
...           ...     ...                ...              ...          ...   
104995         74   Other       West Anthony         Oklahoma          Yes   
104996         37    Male          Mckeetown             Utah          Yes   
104997         51  Female   Lake Lisachester         Nebraska          Yes   
104998        NaN  Female       Heidiborough              NaN          Yes   
104999         45  Female      South Jeffrey    Massachusetts           No   

       visitedlastmonth paymentmethod preferreddoctor diseasedi

## Handeling the missing values

In [ ]:
import re
import numpy as np

In [ ]:
healthcare_data_cleaning['patientage'] = pd.to_numeric(healthcare_data_cleaning['patientage'], errors='coerce')
healthcare_data_cleaning['averageheartrate'] = pd.to_numeric(healthcare_data_cleaning['averageheartrate'], errors='coerce')
healthcare_data_cleaning['averagebp'] = pd.to_numeric(healthcare_data_cleaning['averagebp'], errors='coerce')
healthcare_data_cleaning['heightincm'] = pd.to_numeric(healthcare_data_cleaning['heightincm'], errors='coerce')
healthcare_data_cleaning['weightinkg'] = pd.to_numeric(healthcare_data_cleaning['weightinkg'], errors='coerce')
healthcare_data_cleaning['paymentdue'] = pd.to_numeric(healthcare_data_cleaning['paymentdue'], errors='coerce')
healthcare_data_cleaning['lastvisitdaysago'] = pd.to_numeric(healthcare_data_cleaning['lastvisitdaysago'], errors='coerce')
healthcare_data_cleaning['visitdurationmins'] = pd.to_numeric(healthcare_data_cleaning['visitdurationmins'], errors='coerce')
healthcare_data_cleaning['numberoftests'] = pd.to_numeric(healthcare_data_cleaning['numberoftests'], errors='coerce')
healthcare_data_cleaning['prescriptioncost'] = pd.to_numeric(healthcare_data_cleaning['prescriptioncost'], errors='coerce')

In [ ]:
print(healthcare_data_cleaning['patientage'].value_counts())

18.0     4047
47.0     2555
45.0     2543
44.0     2529
42.0     2518
         ... 
100.0      10
96.0       10
97.0        8
99.0        7
98.0        5
Name: patientage, Length: 83, dtype: int64


In [ ]:
print(healthcare_data_cleaning['gender'].value_counts())


Male      31728
Female    31674
Other     31598
Name: gender, dtype: int64


In [ ]:
print(healthcare_data_cleaning['cityofresidence'].value_counts())

Lake Michael         100
West Michael          82
East Michael          77
Port Michael          74
North Michael         71
                    ... 
Deleontown             1
Lake Hollychester      1
Port Edwardhaven       1
West Pamelaton         1
Mckeetown              1
Name: cityofresidence, Length: 37132, dtype: int64


In [ ]:
print(healthcare_data_cleaning['stateofresidence'].value_counts())


New Mexico        1969
Nevada            1969
Idaho             1965
Massachusetts     1959
California        1952
Ohio              1950
Connecticut       1950
Kentucky          1946
North Carolina    1944
New Jersey        1932
Alaska            1927
Wisconsin         1926
Nebraska          1922
Rhode Island      1918
Minnesota         1918
Wyoming           1917
West Virginia     1916
Arizona           1915
Tennessee         1914
South Carolina    1912
Virginia          1911
North Dakota      1911
Utah              1910
Kansas            1909
Pennsylvania      1908
Missouri          1906
Iowa              1906
Montana           1904
Mississippi       1897
Florida           1893
Louisiana         1893
New York          1891
Illinois          1891
Oregon            1888
Indiana           1886
Vermont           1885
Oklahoma          1876
Texas             1876
Alabama           1875
Michigan          1863
Georgia           1862
Colorado          1859
Delaware          1859
Hawaii     

In [ ]:
print(healthcare_data_cleaning['hasinsurance'].value_counts())


Yes    47597
No     47403
Name: hasinsurance, dtype: int64


In [ ]:
print(healthcare_data_cleaning['visitedlastmonth'].value_counts())


No     47572
Yes    47428
Name: visitedlastmonth, dtype: int64


In [ ]:
print(healthcare_data_cleaning['paymentmethod'].value_counts())


Card         31744
Insurance    31662
Cash         31594
Name: paymentmethod, dtype: int64


In [ ]:
print(healthcare_data_cleaning['preferreddoctor'].value_counts())


Dr. Jones       19153
Dr. Williams    19078
Dr. Johnson     19005
Dr. Smith       18914
Dr. Brown       18850
Name: preferreddoctor, dtype: int64


In [ ]:
print(healthcare_data_cleaning['diseasediagnosed'].value_counts())


Flu         19123
Allergy     19063
None        19021
Cold        18962
Covid-19    18831
Name: diseasediagnosed, dtype: int64


In [ ]:
print(healthcare_data_cleaning['typeofappointment'].value_counts())


Specialist    23841
Follow-Up     23832
Emergency     23737
General       23590
Name: typeofappointment, dtype: int64


In [ ]:
print(healthcare_data_cleaning['averageheartrate'].value_counts())


80.0     3779
81.0     3769
78.0     3764
77.0     3711
82.0     3694
         ... 
114.0      11
116.0       6
120.0       4
117.0       3
119.0       1
Name: averageheartrate, Length: 70, dtype: int64


In [ ]:
print(healthcare_data_cleaning['averagebp'].value_counts())


80.0     2560
110.0    2509
109.0    2509
112.0    2491
111.0    2489
         ... 
82.0      489
137.0     465
81.0      425
138.0     415
139.0     374
Name: averagebp, Length: 61, dtype: int64


In [ ]:
print(healthcare_data_cleaning['heightincm'].value_counts())


174.0    3834
175.0    3767
176.0    3731
177.0    3727
173.0    3702
         ... 
142.0      19
207.0      17
141.0      15
208.0      12
209.0      11
Name: heightincm, Length: 71, dtype: int64


In [ ]:
print(healthcare_data_cleaning['weightinkg'].value_counts())


71.0     2559
40.0     2544
67.0     2532
72.0     2526
69.0     2481
         ... 
134.0       1
130.0       1
132.0       1
128.0       1
135.0       1
Name: weightinkg, Length: 93, dtype: int64


In [ ]:
print(healthcare_data_cleaning['paymentdue'].value_counts())


65.457216     1
299.988500    1
150.184654    1
297.964090    1
29.961924     1
             ..
471.830263    1
456.186160    1
273.771756    1
315.289042    1
5.341231      1
Name: paymentdue, Length: 95000, dtype: int64


In [ ]:
print(healthcare_data_cleaning['lastvisitdaysago'].value_counts())


183.0    3992
182.0    3985
184.0    3927
180.0    3789
185.0    3777
         ... 
223.0       2
140.0       1
220.0       1
221.0       1
222.0       1
Name: lastvisitdaysago, Length: 81, dtype: int64


In [ ]:
print(healthcare_data_cleaning['visitdurationmins'].value_counts())


49.0     931
69.0     924
68.0     917
27.0     914
36.0     906
        ... 
98.0     816
109.0    809
87.0     808
67.0     804
107.0    802
Name: visitdurationmins, Length: 111, dtype: int64


In [ ]:
print(healthcare_data_cleaning['numberoftests'].value_counts())


4.0     23795
3.0     20400
5.0     19094
2.0     11541
6.0     10566
7.0      3993
1.0      3925
8.0       979
0.0       537
9.0       160
10.0       10
Name: numberoftests, dtype: int64


In [ ]:
print(healthcare_data_cleaning['prescriptioncost'].value_counts())


15.717517     1
65.992965     1
134.851044    1
67.139600     1
4.058180      1
             ..
86.083105     1
50.488526     1
78.533369     1
124.019842    1
124.408341    1
Name: prescriptioncost, Length: 95000, dtype: int64


In [ ]:
total_missing_values = healthcare_data_cleaning.isnull().sum()
total_missing_values

patient_age              5249
gender                   5261
city_of_residence        5227
state_of_residence       5248
has_insurance            5246
visited_last_month       5245
payment_method           5202
preferred_doctor         5247
disease_diagnosed        5271
medication_prescribed    5262
type_of_appointment      5250
average_heart_rate       5278
average_bp               5255
height_in_cm             5244
weight_in_kg             5264
payment_due_$            5229
last_visit_days_ago      5259
visit_duration_mins      5276
number_of_tests          5272
prescription_cost_$      5216
dtype: int64

In [ ]:
healthcare_data_cleaning= healthcare_data_cleaning.replace('missing', 'N/A')
healthcare_data_cleaning

patient_age  gender  city_of_residence state_of_residence  \
0               45   Other         West Brian           Kentucky   
1               25  Female   East Jocelynfurt           Colorado   
2               51   Other  South Lindseyland               Ohio   
3               18  Female         Taylorfort       Pennsylvania   
4               48  Female          Birdmouth            Montana   
...            ...     ...                ...                ...   
104995          74   Other       West Anthony           Oklahoma   
104996          37    Male          Mckeetown               Utah   
104997          51  Female   Lake Lisachester           Nebraska   
104998         NaN  Female       Heidiborough                NaN   
104999          45  Female      South Jeffrey      Massachusetts   

       has_insurance visited_last_month payment_method preferred_doctor  \
0                Yes                Yes      Insurance     Dr. Williams   
1                 No                Yes           Card      Dr. Johnson   
2                Yes                 No           Cash     Dr. Williams   
3                Yes                NaN           Cash     Dr. Williams   
4                 No                 No      Insurance        Dr. Brown   
...              ...                ...            ...              ...   
104995           Yes                 No           Card     Dr. Williams   
104996           Yes                 No           Cash        Dr. Jones   
104997           Yes                Yes      Insurance     Dr. Williams   
104998           Yes                 No           Cash        Dr. Smith   
104999            No                Yes           Cash        Dr. Brown   

       disease_diagnosed medication_prescribed type_of_appointment  \
0                    Flu                 Med_E             General   
1                    Flu                 Med_B          Specialist   
2               Covid-19                 Med_D             General   
3                    Flu                 Med_A             General   
4                   None                 Med_C          Specialist   
...                  ...                   ...                 ...   
104995          Covid-19                 Med_E          Specialist   
104996          Covid-19                 Med_C             General   
104997           Allergy                 Med_D           Follow-Up   
104998          Covid-19                 Med_A           Follow-Up   
104999           Allergy                 Med_D             General   

       average_heart_rate average_bp height_in_cm weight_in_kg  \
0                      66        111          183           70   
1                      59        NaN          174           58   
2                      79        119          161           56   
3                      99        115          171           52   
4                      93         89          186           87   
...                   ...        ...          ...          ...   
104995                 65        112          169           61   
104996                NaN        111          165           67   
104997                 89        118          179           71   
104998                 74        123          167           75   
104999                 77        140          167           62   

             payment_due_$ last_visit_days_ago visit_duration_mins  \
0        65.45721578126224                 193                  92   
1       430.68365678679174                 195                 NaN   
2        315.0709305262176                 NaN                  37   
3        320.2998987723972                 175                 108   
4        264.2147372473319                 188                 113   
...                    ...                 ...                 ...   
104995  309.16444606680665                 183                  10   
104996                 NaN                 191                  47   
104997   320.830418003390

## Cleaning column names

In [ ]:
healthcare_data_cleaning.columns

Index(['patient_age', 'gender', 'city_of_residence', 'state_of_residence',
       'has_insurance', 'visited_last_month', 'payment_method',
       'preferred_doctor', 'disease_diagnosed', 'medication_prescribed',
       'type_of_appointment', 'average_heart_rate', 'average_bp',
       'height_in_cm', 'weight_in_kg', 'payment_due_$', 'last_visit_days_ago',
       'visit_duration_mins', 'number_of_tests', 'prescription_cost_$'],
      dtype='object')

In [ ]:
healthcare_data_cleaning.columns = healthcare_data_cleaning.columns.str.strip().str.lower().str.replace(' ', '_').str.replace('(', '').str.replace(')', '')
healthcare_data_cleaning

<ipython-input-10-bde3266830c7>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  healthcare_data_cleaning.columns = healthcare_data_cleaning.columns.str.strip().str.lower().str.replace(' ', '_').str.replace('(', '').str.replace(')', '')
<ipython-input-10-bde3266830c7>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  healthcare_data_cleaning.columns = healthcare_data_cleaning.columns.str.strip().str.lower().str.replace(' ', '_').str.replace('(', '').str.replace(')', '')


patient_age  gender  city_of_residence state_of_residence  \
0               45   Other         West Brian           Kentucky   
1               25  Female   East Jocelynfurt           Colorado   
2               51   Other  South Lindseyland               Ohio   
3               18  Female         Taylorfort       Pennsylvania   
4               48  Female          Birdmouth            Montana   
...            ...     ...                ...                ...   
104995          74   Other       West Anthony           Oklahoma   
104996          37    Male          Mckeetown               Utah   
104997          51  Female   Lake Lisachester           Nebraska   
104998         NaN  Female       Heidiborough                NaN   
104999          45  Female      South Jeffrey      Massachusetts   

       has_insurance visited_last_month payment_method preferred_doctor  \
0                Yes                Yes      Insurance     Dr. Williams   
1                 No                Yes           Card      Dr. Johnson   
2                Yes                 No           Cash     Dr. Williams   
3                Yes                NaN           Cash     Dr. Williams   
4                 No                 No      Insurance        Dr. Brown   
...              ...                ...            ...              ...   
104995           Yes                 No           Card     Dr. Williams   
104996           Yes                 No           Cash        Dr. Jones   
104997           Yes                Yes      Insurance     Dr. Williams   
104998           Yes                 No           Cash        Dr. Smith   
104999            No                Yes           Cash        Dr. Brown   

       disease_diagnosed medication_prescribed type_of_appointment  \
0                    Flu                 Med_E             General   
1                    Flu                 Med_B          Specialist   
2               Covid-19                 Med_D             General   
3                    Flu                 Med_A             General   
4                   None                 Med_C          Specialist   
...                  ...                   ...                 ...   
104995          Covid-19                 Med_E          Specialist   
104996          Covid-19                 Med_C             General   
104997           Allergy                 Med_D           Follow-Up   
104998          Covid-19                 Med_A           Follow-Up   
104999           Allergy                 Med_D             General   

       average_heart_rate average_bp height_in_cm weight_in_kg  \
0                      66        111          183           70   
1                      59        NaN          174           58   
2                      79        119          161           56   
3                      99        115          171           52   
4                      93         89          186           87   
...                   ...        ...          ...          ...   
104995                 65        112          169           61   
104996                NaN        111          165           67   
104997                 89        118          179           71   
104998                 74        123          167           75   
104999                 77        140          167           62   

             payment_due_$ last_visit_days_ago visit_duration_mins  \
0        65.45721578126224                 193                  92   
1       430.68365678679174                 195                 NaN   
2        315.0709305262176                 NaN                  37   
3        320.2998987723972                 175                 108   
4        264.2147372473319                 188                 113   
...                    ...                 ...                 ...   
104995  309.16444606680665                 183                  10   
104996                 NaN                 191                  47   
104997   320.830418003390

# 2. Data Transformation:

## Creating New Column Names Based on Existing Ones




In [ ]:
column_mapping = {
    'patient_age': 'Age',
    'city_of_residence': 'City',
    'state_of_residence': 'State',
    'has_insurace': 'Insurance',
    'payment_method': 'Payment',
    'preferred_doctor': 'Preferred_Dr',
    'disease_diagnosed': 'Disease_Diagnosis',
    'medication_prescribed': 'Prescription',
    'height_in_cm': 'Height_cm',
    'weight_in_kg': 'Weight_kg'
}
healthcare_data_cleaning.rename(columns=column_mapping, inplace=True)
healthcare_data_cleaning

Patient Age  Gender  City of Residence State of Residence  \
0               45   Other         West Brian           Kentucky   
1               25  Female   East Jocelynfurt           Colorado   
2               51   Other  South Lindseyland               Ohio   
3               18  Female         Taylorfort       Pennsylvania   
4               48  Female          Birdmouth            Montana   
...            ...     ...                ...                ...   
104995          74   Other       West Anthony           Oklahoma   
104996          37    Male          Mckeetown               Utah   
104997          51  Female   Lake Lisachester           Nebraska   
104998         NaN  Female       Heidiborough                NaN   
104999          45  Female      South Jeffrey      Massachusetts   

       Has Insurance Visited Last Month Payment Method Preferred Doctor  \
0                Yes                Yes      Insurance     Dr. Williams   
1                 No                Yes           Card      Dr. Johnson   
2                Yes                 No           Cash     Dr. Williams   
3                Yes                NaN           Cash     Dr. Williams   
4                 No                 No      Insurance        Dr. Brown   
...              ...                ...            ...              ...   
104995           Yes                 No           Card     Dr. Williams   
104996           Yes                 No           Cash        Dr. Jones   
104997           Yes                Yes      Insurance     Dr. Williams   
104998           Yes                 No           Cash        Dr. Smith   
104999            No                Yes           Cash        Dr. Brown   

       Disease Diagnosed Medication Prescribed Type of Appointment  \
0                    Flu                 Med_E             General   
1                    Flu                 Med_B          Specialist   
2               Covid-19                 Med_D             General   
3                    Flu                 Med_A             General   
4                   None                 Med_C          Specialist   
...                  ...                   ...                 ...   
104995          Covid-19                 Med_E          Specialist   
104996          Covid-19                 Med_C             General   
104997           Allergy                 Med_D           Follow-Up   
104998          Covid-19                 Med_A           Follow-Up   
104999           Allergy                 Med_D             General   

       Average Heart Rate Average BP Height (in cm) Weight (in kg)  \
0                      66        111            183             70   
1                      59        NaN            174             58   
2                      79        119            161             56   
3                      99        115            171             52   
4                      93         89            186             87   
...                   ...        ...            ...            ...   
104995                 65        112            169             61   
104996                NaN        111            165             67   
104997                 89        118            179             71   
104998                 74        123            167             75   
104999                 77        140            167             62   

           Payment Due ($) Last Visit (days ago) Visit Duration (mins)  \
0        65.45721578126224                   193                    92   
1       430.68365678679174                   195                   NaN   
2        315.0709305262176                   NaN                    37   
3        320.2998987723972                   175                   108   
4        264.2147372473319                   188                   113   
...                    ...                   ...                   ...   
104995  309.16444606680665                   183                    10   
104996         

## Aggreating Data Using Groupby and Compute Summary Statistics

In [ ]:
print(healthcare_data_cleaning['visitedlastmonth'].dtypes)

object


In [ ]:
print(healthcare_data_cleaning['paymentdue'].dtypes)

float64


In [ ]:
print(healthcare_data_cleaning['diseasediagnosed'].dtypes)

object


In [ ]:
healthcare_data_cleaning['diseasediagnosed'] = pd.to_numeric(healthcare_data_cleaning['diseasediagnosed'], errors='coerce')

healthcare_data_cleaning['medicationprescribed'].value_counts()

Med_C    19149
Med_D    19040
Med_B    19003
Med_A    18957
Med_E    18851
Name: medicationprescribed, dtype: int64

In [ ]:
grouped_data = healthcare_data_cleaning.groupby('medicationprescribed').agg({
    'diseasediagnosed': 'mean'
})
grouped_data

diseasediagnosed
medicationprescribed                  
Med_A                       249.285961
Med_B                       249.196270
Med_C                       248.356487
Med_D                       251.265073
Med_E                       249.905104

In [ ]:
the_disease_diagnosis = healthcare_data_cleaning[healthcare_data_cleaning['diseasediagnosed'] == 'Flu']
the_disease_diagnosis.groupby(['diseasediagnosed']).sum()


<ipython-input-45-beb358b7284a>:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  the_disease_diagnosis.groupby(['diseasediagnosed']).sum()


patientage  averageheartrate  averagebp  heightincm  \
diseasediagnosed                                                        
Flu                 811046.0         1442699.0  1987959.0   3168630.0   

                  weightinkg    paymentdue  lastvisitdaysago  \
diseasediagnosed                                               
Flu                1262238.0  4.500347e+06         3313991.0   

                  visitdurationmins  numberoftests  prescriptioncost  
diseasediagnosed                                                      
Flu                       1179560.0        72505.0      1.359500e+06

In [ ]:
pivot_table_data = healthcare_data_cleaning.pivot_table(index='cityofresidence', columns='diseasediagnosed', values='paymentdue', aggfunc='mean')
pivot_table_data

diseasediagnosed     Allergy        Cold    Covid-19         Flu        None
cityofresidence                                                             
Aaronberg          53.435276         NaN         NaN  200.185919         NaN
Aaronborough      174.035005  425.948239  417.036905         NaN  101.336997
Aaronburgh        302.079018  163.291124  191.207121  426.693173  348.700014
Aaronbury                NaN         NaN  217.983568  183.814863         NaN
Aaronchester      331.706527         NaN  144.269185         NaN  290.556463
...                      ...         ...         ...         ...         ...
Zoeburgh                 NaN         NaN         NaN  458.810402         NaN
Zoechester               NaN  142.003169         NaN         NaN         NaN
Zunigaland               NaN         NaN         NaN  346.871637         NaN
Zunigamouth              NaN         NaN         NaN  379.164892         NaN
Zunigaport               NaN  265.356088         NaN         NaN         NaN

[35083 rows x 5 columns]

In [ ]:
pivot_table = healthcare_data_cleaning.pivot_table(index='gender', columns='hasinsurance', values='paymentdue', aggfunc='mean')
pivot_table

hasinsurance          No         Yes
gender                              
Female        248.505546  249.048480
Male          251.145420  249.736330
Other         248.469690  250.898797

# 3. Introducion to Alternative Libraries

## Loading Polar Package

In [ ]:
import polars as pl

In [ ]:
healthcare_data_cleaning = pl.read_csv('https://raw.githubusercontent.com/nimratmann/datasci_2_manipulation/main/data/healthcare%20data%20cleaning.csv')
healthcare_data_cleaning


shape: (105_000, 20)
┌───────────┬────────┬────────────┬────────────┬───┬────────────┬────────────┬────────────┬────────────┐
│ Patient   ┆ Gender ┆ City of    ┆ State of   ┆ … ┆ Last Visit ┆ Visit      ┆ Number of  ┆ Prescripti │
│ Age       ┆ ---    ┆ Residence  ┆ Residence  ┆   ┆ (days ago) ┆ Duration   ┆ Tests      ┆ on Cost    │
│ ---       ┆ str    ┆ ---        ┆ ---        ┆   ┆ ---        ┆ (mins)     ┆ ---        ┆ ($)        │
│ str       ┆        ┆ str        ┆ str        ┆   ┆ str        ┆ ---        ┆ str        ┆ ---        │
│           ┆        ┆            ┆            ┆   ┆            ┆ str        ┆            ┆ str        │
╞═══════════╪════════╪════════════╪════════════╪═══╪════════════╪════════════╪════════════╪════════════╡
│ 45        ┆ Other  ┆ West Brian ┆ Kentucky   ┆ … ┆ 193        ┆ 92         ┆ 5          ┆ 15.7175173 │
│           ┆        ┆            ┆            ┆   ┆            ┆            ┆            ┆ 5532889    │
│ 25        ┆ Female ┆ East Jocel ┆ Colorado   ┆ … ┆ 195        ┆ missing    ┆ 4          ┆ 80.7864728 │
│           ┆        ┆ ynfurt     ┆            ┆   ┆            ┆            ┆            ┆ 4463952    │
│ 51        ┆ Other  ┆ South Lind ┆ Ohio       ┆ … ┆ missing    ┆ 37         ┆ 2          ┆ 64.3921393 │
│           ┆        ┆ seyland    ┆            ┆   ┆            ┆            ┆            ┆ 009105     │
│ 18        ┆ Female ┆ Taylorfort ┆ Pennsylvan ┆ … ┆ 175        ┆ 108        ┆ 2          ┆ 8.87285921 │
│           ┆        ┆            ┆ ia         ┆   ┆            ┆            ┆            ┆ 5315316    │
│ …         ┆ …      ┆ …          ┆ …          ┆ … ┆ …          ┆ …          ┆ …          ┆ …          │
│ 37        ┆ Male   ┆ Mckeetown  ┆ Utah       ┆ … ┆ 191        ┆ 47         ┆ 2          ┆ 113.867640 │
│           ┆        ┆            ┆            ┆   ┆            ┆            ┆            ┆ 25370326   │
│ 51        ┆ Female ┆ Lake Lisac ┆ Nebraska   ┆ … ┆ 172        ┆ 39         ┆ 3          ┆ 35.6224202 │
│           ┆        ┆ hester     ┆            ┆   ┆            ┆            ┆            ┆ 32475356   │
│ missing   ┆ Female ┆ Heidiborou ┆ missing    ┆ … ┆ 180        ┆ 79         ┆ missing    ┆ 124.408340 │
│           ┆        ┆ gh         ┆            ┆   ┆            ┆            ┆            ┆ 9893742    │
│ 45        ┆ Female ┆ South      ┆ Massachuse ┆ … ┆ 167        ┆ 81         ┆ missing    ┆ 87.9106014 │
│           ┆        ┆ Jeffrey    ┆ tts        ┆   ┆            ┆            ┆            ┆ 9353706    │
└───────────┴────────┴────────────┴────────────┴───┴────────────┴────────────┴────────────┴────────────┘

## Polars Loading Time

In [ ]:
import time
dataset = 'health_data_cleaning.csv'
start_time = time.time()
polars_healthcare_data_cleaning = pd.read_csv('https://raw.githubusercontent.com/nimratmann/datasci_2_manipulation/main/data/healthcare%20data%20cleaning.csv')
polars_filtered = polars_healthcare_data_cleaning[polars_healthcare_data_cleaning['Gender'] == 'Male']
polars_average_age = polars_filtered['Average BP'].sum()
polars_time = time.time() - start_time

In [60]:
print(f"Polars Load Time: {polars_time:.4f} seconds")

Polars Load Time: 2.0262 seconds


In [61]:
import pandas as pd
import time

healthcare_data_cleaning = pd.read_csv('https://raw.githubusercontent.com/nimratmann/datasci_2_manipulation/main/data/healthcare%20data%20cleaning.csv')

print(f"Pandas load time: {load_time_seconds:.4f} seconds")

Pandas load time: 0.0000 seconds


### Observations Based on Loading Times

##### In this case, Polars took 2.0262 seconds to load the same CSV data, which is noticeably longer compared to Pandas. The reason for the longer load time might be due to factors like the initial setup, data transformation, or other optimizations specific to Polars.

##### When loading data with Pandas, it often uses low-level libraries like NumPy, which can be optimized for speed. In this case, Pandas loaded the CSV data so quickly that it was perceived as taking "no time" or a very minimal amount of time (0.0000 seconds).